File: App/BNKScopeField/Theme.swift
import SwiftUI

/// The bnkscope web UI's dark tokens, carried over unchanged so the two read as
/// one product. The one addition is `live`: the logo's red, reserved here for
/// telemetry that is actually streaming. Blue stays what it is on the web —
/// selection and interactive chrome — so "selected" and "live" never look alike.
enum Theme {
    static let bg        = Color(hex: 0x111217)
    static let card      = Color(hex: 0x161A1F)
    static let border    = Color(hex: 0x303740)
    static let fg        = Color(hex: 0xCCCCDC)
    static let muted     = Color(hex: 0x7C7C9C)
    static let faint     = Color(hex: 0x5C6270)
    static let secondary = Color(hex: 0x1E2028)
    static let mutedBg   = Color(hex: 0x22252B)

    static let primary = Color(hex: 0x2563EB)
    static let ember   = Color(hex: 0xFF6A00)
    static let deep    = Color(hex: 0xE4002B)
    /// The logo's red. It belongs to the mark, and to nothing else.
    ///
    /// It was the "streaming" colour until someone pointed out that a
    /// troubleshooting tool showing its healthiest state in red is fighting
    /// every other red on the screen — including STALLED, which at #EF4444 was
    /// close enough to be mistaken for it at a glance. Broadcast convention says
    /// a live indicator is red; a monitoring convention says red is trouble, and
    /// in here the monitoring one wins.
    static let brand   = Color(hex: 0xFF3355)

    static let ok      = Color(hex: 0x10B981)
    static let warn    = Color(hex: 0xF59E0B)
    static let bad     = Color(hex: 0xEF4444)

    /// Series colours, assigned in this fixed order and never cycled.
    ///
    /// These are the web UI's chart hues stepped down for a dark surface — the
    /// originals sit above the lightness band that reads well on #161A1F. The
    /// three status hues are deliberately absent: amber, red and green mean
    /// something here, and a series that borrows one lies about its state.
    static let series: [Color] = [
        Color(hex: 0x3B82F6),   // blue
        Color(hex: 0x0D9488),   // teal
        Color(hex: 0x8B5CF6),   // violet
        Color(hex: 0xDB2777),   // pink
        Color(hex: 0x0891B2),   // cyan
    ]

    static func seriesColor(_ index: Int) -> Color {
        index < series.count ? series[index] : muted
    }

    static let mono = Font.system(.body, design: .monospaced)
    static func mono(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
        .system(size: size, weight: weight, design: .monospaced)
    }
}

extension Color {
    init(hex: UInt32) {
        self.init(.sRGB,
                  red:   Double((hex >> 16) & 0xFF) / 255,
                  green: Double((hex >> 8)  & 0xFF) / 255,
                  blue:  Double(hex & 0xFF) / 255,
                  opacity: 1)
    }
}

/// The bnkscope mark: a heptagonal scope with a waveform across it. Traced from
/// `frontend-v2/public/icons/bnkscope-small.svg` so the two stay the same shape.
struct BNKMark: View {
    var size: CGFloat = 28

    private static let bezel = Path { p in
        p.move(to: CGPoint(x: 32, y: 9.6))
        for pt in [CGPoint(x: 49.98, y: 18.26), CGPoint(x: 54.42, y: 37.72), CGPoint(x: 41.98, y: 53.32),
                   CGPoint(x: 22.02, y: 53.32), CGPoint(x: 9.58, y: 37.72), CGPoint(x: 14.02, y: 18.26)] {
            p.addLine(to: pt)
        }
        p.closeSubpath()
    }

    private static let beam = Path { p in
        let pts: [CGPoint] = [(15, 33), (18, 32.92), (19.5, 30.7), (21, 26.58), (22.5, 26.39),
                              (24.5, 28.48), (27, 32.82), (29, 35.74), (31, 37.01), (33, 36.44),
                              (35, 34.63), (37, 32.53), (39, 31.03), (41, 30.6), (43, 31.2),
                              (45, 32.39), (47, 33.59), (48, 34.03)].map { CGPoint(x: $0.0, y: $0.1) }
        p.move(to: pts[0])
        for pt in pts.dropFirst() { p.addLine(to: pt) }
    }

    var body: some View {
        Canvas { ctx, _ in
            ctx.fill(RoundedRectangle(cornerRadius: 14).path(in: CGRect(x: 2, y: 2, width: 60, height: 60)),
                     with: .color(Color(hex: 0x15181D)))
            ctx.fill(Self.bezel, with: .color(Color(hex: 0x0A0C10)))
            ctx.clip(to: Self.bezel)
            ctx.stroke(Self.beam,
                       with: .linearGradient(Gradient(colors: [Theme.brand, Theme.brand, Theme.ember]),
                                             startPoint: CGPoint(x: 12, y: 0), endPoint: CGPoint(x: 52, y: 0)),
                       style: StrokeStyle(lineWidth: 4, lineCap: .round, lineJoin: .round))
        }
        .frame(width: 64, height: 64)
        .overlay {
            Self.bezel.stroke(Theme.deep, style: StrokeStyle(lineWidth: 2.6, lineJoin: .round))
                .frame(width: 64, height: 64)
        }
        .scaleEffect(size / 64)
        .frame(width: size, height: size)
    }
}

File: dart/bnkscope_field/lib/theme.dart
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

/// The bnkscope web UI's dark tokens, carried over unchanged so the two
/// read as one product.
///
/// Colour is a monitoring convention, not broadcast: green means data is
/// flowing, red means it is not. The brand red belongs to the mark and
/// nothing else.
class Tokens {
  static const bg = Color(0xFF111217);
  static const card = Color(0xFF161A1F);
  static const border = Color(0xFF303740);
  static const fg = Color(0xFFCCCCDC);
  static const muted = Color(0xFF7C7C9C);
  static const faint = Color(0xFF5C6270);
  static const secondary = Color(0xFF1E2028);
  static const mutedBg = Color(0xFF22252B);

  static const primary = Color(0xFF2563EB);
  static const ember = Color(0xFFFF6A00);
  static const deep = Color(0xFFE4002B);

  /// The logo's red. It belongs to the mark, and to nothing else.
  static const brand = Color(0xFFFF3355);

  static const ok = Color(0xFF10B981);
  static const warn = Color(0xFFF59E0B);
  static const bad = Color(0xFFEF4444);

  /// The active row's text, a touch brighter than [fg].
  static const activeFg = Color(0xFFDFE7F7);

  /// The dot of a cluster that cannot be reached.
  static const deadDot = Color(0xFF4B515E);

  /// Series colours, assigned in this fixed order and never cycled. The
  /// three status hues are deliberately absent: amber, red and green mean
  /// something here, and a series that borrows one lies about its state.
  static const series = [
    Color(0xFF3B82F6), // blue
    Color(0xFF0D9488), // teal
    Color(0xFF8B5CF6), // violet
    Color(0xFFDB2777), // pink
    Color(0xFF0891B2), // cyan
  ];

  static Color seriesColor(int index) => index < series.length ? series[index] : muted;

  static const monoFamily = 'Menlo';
  static const monoFallback = ['SF Mono', 'Consolas', 'DejaVu Sans Mono', 'Roboto Mono', 'monospace'];

  static TextStyle mono(double size, {FontWeight weight = FontWeight.w400, Color color = fg}) => TextStyle(
      fontFamily: monoFamily, fontFamilyFallback: monoFallback, fontSize: size, fontWeight: weight, color: color);

  static TextStyle text(double size, {FontWeight weight = FontWeight.w400, Color color = fg}) =>
      TextStyle(fontSize: size, fontWeight: weight, color: color);

  static ThemeData get theme {
    final base = ThemeData(brightness: Brightness.dark, useMaterial3: true);
    return base.copyWith(
      scaffoldBackgroundColor: bg,
      canvasColor: bg,
      colorScheme: const ColorScheme.dark(
        primary: primary,
        onPrimary: Colors.white,
        surface: card,
        onSurface: fg,
        error: bad,
        outline: border,
      ),
      dividerColor: border,
      textTheme: base.textTheme.apply(bodyColor: fg, displayColor: fg),
      filledButtonTheme: FilledButtonThemeData(
        style: FilledButton.styleFrom(
          backgroundColor: primary,
          foregroundColor: Colors.white,
          visualDensity: VisualDensity.compact,
          padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
          textStyle: const TextStyle(fontSize: 12.5, fontWeight: FontWeight.w600),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(7)),
        ),
      ),
      outlinedButtonTheme: OutlinedButtonThemeData(
        style: OutlinedButton.styleFrom(
          foregroundColor: fg,
          backgroundColor: secondary,
          side: const BorderSide(color: border),
          visualDensity: VisualDensity.compact,
          padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
          textStyle: const TextStyle(fontSize: 12.5, fontWeight: FontWeight.w600),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(7)),
        ),
      ),
      textButtonTheme: TextButtonThemeData(
        style: TextButton.styleFrom(foregroundColor: fg, visualDensity: VisualDensity.compact),
      ),
      dialogTheme: DialogThemeData(
        backgroundColor: card,
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12), side: const BorderSide(color: border)),
        titleTextStyle: text(16, weight: FontWeight.w600),
        contentTextStyle: text(13, color: muted),
      ),
      drawerTheme: const DrawerThemeData(backgroundColor: card),
      progressIndicatorTheme: const ProgressIndicatorThemeData(color: primary),
      splashFactory: NoSplash.splashFactory,
    );
  }
}

/// The bnkscope mark: a heptagonal scope with a waveform across it. Traced
/// from `frontend-v2/public/icons/bnkscope-small.svg` so the two stay the
/// same shape.
class BNKMark extends StatelessWidget {
  final double size;
  const BNKMark({super.key, this.size = 28});

  @override
  Widget build(BuildContext context) =>
      SizedBox(width: size, height: size, child: CustomPaint(painter: _MarkPainter()));
}

class _MarkPainter extends CustomPainter {
  static final bezel = Path()
    ..moveTo(32, 9.6)
    ..lineTo(49.98, 18.26)
    ..lineTo(54.42, 37.72)
    ..lineTo(41.98, 53.32)
    ..lineTo(22.02, 53.32)
    ..lineTo(9.58, 37.72)
    ..lineTo(14.02, 18.26)
    ..close();

  static final beam = () {
    const pts = [
      (15.0, 33.0), (18.0, 32.92), (19.5, 30.7), (21.0, 26.58), (22.5, 26.39), (24.5, 28.48), (27.0, 32.82),
      (29.0, 35.74), (31.0, 37.01), (33.0, 36.44), (35.0, 34.63), (37.0, 32.53), (39.0, 31.03), (41.0, 30.6),
      (43.0, 31.2), (45.0, 32.39), (47.0, 33.59), (48.0, 34.03),
    ];
    final p = Path()..moveTo(pts[0].$1, pts[0].$2);
    for (final pt in pts.skip(1)) {
      p.lineTo(pt.$1, pt.$2);
    }
    return p;
  }();

  @override
  void paint(Canvas canvas, Size size) {
    canvas.scale(size.width / 64, size.height / 64);
    canvas.drawRRect(RRect.fromRectAndRadius(const Rect.fromLTWH(2, 2, 60, 60), const Radius.circular(14)),
        Paint()..color = const Color(0xFF15181D));
    canvas.drawPath(bezel, Paint()..color = const Color(0xFF0A0C10));
    canvas.save();
    canvas.clipPath(bezel);
    canvas.drawPath(
        beam,
        Paint()
          ..style = PaintingStyle.stroke
          ..strokeWidth = 4
          ..strokeCap = StrokeCap.round
          ..strokeJoin = StrokeJoin.round
          ..shader = ui.Gradient.linear(const Offset(12, 0), const Offset(52, 0),
              const [Tokens.brand, Tokens.brand, Tokens.ember], const [0, 0.5, 1]));
    canvas.restore();
    canvas.drawPath(
        bezel,
        Paint()
          ..style = PaintingStyle.stroke
          ..strokeWidth = 2.6
          ..strokeJoin = StrokeJoin.round
          ..color = Tokens.deep);
  }

  @override
  bool shouldRepaint(_MarkPainter old) => false;
}
